[nvbug/5280806][fix] Fix 2 model spec decode flow

mikeiovine opened the pull request:
# [fix] Fix 2 model spec decode flow

## Description

Fix a small issue introduced by the `feat/llama4` integration (https://github.com/NVIDIA/TensorRT-LLM/pull/4739).

There was a bad merge:
* In https://github.com/NVIDIA/TensorRT-LLM/pull/3686, we did some code cleanup on the 2 engine path. This line was confusing because of the hardcoded layer index:
```
spec_metadata.maybe_capture_hidden_states(1, hidden_states_to_save)
```
The fix was to capture the hidden states in the `midlayer`, which makes everything consistent with how hidden states are captured in the target model.
* The feat/llama4 branch added these lines back by mistake. They don't work any more due to an API change.

The one model flow doesn't use this code path, so it was probably untested.

Also fixes another small issue introduced by https://github.com/NVIDIA/TensorRT-LLM/pull/4379. The `prompt_length` was not being set correctly for `extend_ctx` cases.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->
Re-enabled EAGLE3 2 model tests in L0.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

mikeiovine committed:
[fix] Fix 2 model spec decode flow

Signed-off-by: Mike Iovine <6158008+mikeiovine@users.noreply.github.com>

mikeiovine commented:
/bot run

mikeiovine commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7233](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7233/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5238](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5238/) completed with status: 'FAILURE'

mikeiovine commented:
Fixed one more issue (`extend_ctx` method using `isinstance` instead of `issubclass`, causing the prefill kernels to be used instead of the new generation kernels on Blackwell).

Also uncovered one more bug. Added a workaround in the unit tests and opened https://nvbugspro.nvidia.com/bug/5314469 to follow up.

mikeiovine commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7240](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7240/) [ run ] triggered by Bot

mikeiovine committed:
Merge branch 'main' into fix-2-model

mikeiovine commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7357](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7357/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7357](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7357/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5333](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5333/) completed with status: 'FAILURE'

mikeiovine commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7376](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7376/) [ run ] triggered by Bot

mikeiovine committed:
Merge branch 'main' into fix-2-model

tensorrt-cicd commented:
[PR_Github #7384](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7384/) [ run ] triggered by Bot

mikeiovine committed:
Merge branch 'main' into fix-2-model

mikeiovine commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7391/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7391/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5362](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5362/) completed with status: 'FAILURE'

mikeiovine commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #7395](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7395/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7395](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7395/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5366](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5366/) completed with status: 'FAILURE'

mikeiovine committed:
Merge branch 'main' into fix-2-model

tensorrt-cicd commented:
[PR_Github #7530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7530/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7530/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5462](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5462/) completed with status: 'FAILURE'

mikeiovine committed:
Merge branch 'main' into fix-2-model

tensorrt-cicd commented:
[PR_Github #7742](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7742/) [ run ] triggered by Bot

mikeiovine committed:
Merge branch 'main' into fix-2-model

tensorrt-cicd commented:
[PR_Github #7792](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7792/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7792](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7792/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5643](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5643/) completed with status: 'FAILURE'

mikeiovine committed:
Merge branch 'main' into fix-2-model

mikeiovine commented:
/bot run --disable-fail-fast

mikeiovine committed:
Merge branch 'main' into fix-2-model

mikeiovine commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #7936](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7936/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7915](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7915/) [ run ] completed with state `ABORTED`

mikeiovine committed:
Merge branch 'main' into fix-2-model
